homeassistant/components/update/__init__.py

```diff
@@ -0,0 +1,257 @@
+"""Support for Update."""
+from __future__ import annotations
+
+import asyncio
+import dataclasses
+import logging
+from typing import Any, Protocol
+
+import async_timeout
+import voluptuous as vol
+
+from homeassistant.components import websocket_api
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import integration_platform, storage
+from homeassistant.helpers.typing import ConfigType
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = "update"
+INFO_CALLBACK_TIMEOUT = 5
+STORAGE_VERSION = 1
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the Update integration."""
+    hass.data[DOMAIN] = UpdateManager(hass=hass)
+    websocket_api.async_register_command(hass, handle_info)
+    websocket_api.async_register_command(hass, handle_update)
+    websocket_api.async_register_command(hass, handle_skip)
+    return True
+
+
+@websocket_api.websocket_command({vol.Required("type"): "update/info"})
+@websocket_api.async_response
+async def handle_info(
+    hass: HomeAssistant,
+    connection: websocket_api.ActiveConnection,
+    msg: dict[str, Any],
+) -> None:
+    """Get pending updates from all platforms."""
+    manager: UpdateManager = hass.data[DOMAIN]
+    updates = await manager.gather_updates()
+    connection.send_result(msg["id"], updates)
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "update/skip",
+        vol.Required("domain"): str,
+        vol.Required("identifier"): str,
+        vol.Required("version"): str,
+    }
+)
+@websocket_api.async_response
+async def handle_skip(
+    hass: HomeAssistant,
+    connection: websocket_api.ActiveConnection,
+    msg: dict[str, Any],
+) -> None:
+    """Skip an update."""
+    manager: UpdateManager = hass.data[DOMAIN]
+
+    if not await manager.domain_is_valid(msg["domain"]):
+        connection.send_error(
+            msg["id"], websocket_api.ERR_NOT_FOUND, "Domain not supported"
+        )
+        return
+
+    manager.skip_update(msg["domain"], msg["identifier"], msg["version"])
```

Comment:
What if identifier or version is no longer valid ? We only validate domain. Would it help to drop `domain_is_valid` and just raise some exception if something is off?

Reply:
Since we do not store updates anymore, that would mean that on each call info/update/skip we would need to gather all updates, then iterate over all updates for the domain to ensure the identifier and version are present.
That's expensive, and this I would say is a rare edge case, and if needed is something that the individual update platforms should handle, they get the identifier and version, and can raise or just return early if the update no longer exists.